cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+	// mask is the mask used to select which signers participated in this round
+	// or not. All code regarding the mask is directly inspired from
+	// github.com/bford/golang-x-crypto/ed25519/cosi code.
+	mask *CosiMask
+	// the longterm private key we use during the rounds
+	private abstract.Secret
+	// the message being co-signed
+	message []byte
+	// timestamp of when the announcement is done (i.e. timestamp of the four
+	// phases)
+	timestamp int64
+	// random is our own secret that we wish to commit during the commitment phase.
+	random abstract.Secret
+	// commitment is our own commitment
+	commitment abstract.Point
+	// V_hat is the aggregated commit (our own + the children's)
+	aggregateCommitment abstract.Point
+	// globalComitment is the global commit of all participants
+	//passed down in the challenge phase
+	globalCommitment abstract.Point
+	// challenge holds the challenge for this round
+	challenge abstract.Secret
+	// response is our own computed response
+	response abstract.Secret
+	// aggregateResponses is the aggregated response from the children + our own
+	aggregateResponse abstract.Secret
+}
+
+// NewCosi returns a new Cosi struct given the suite, the longterm secret, and
+// the list of public keys. If some signers were not to be participating, you
+// have to set the mask using `SetMask` method. By default, all participants are
+// designated as participating.
+func NewCosi(suite abstract.Suite, private abstract.Secret, publics []abstract.Point) *Cosi {
+	cos := &Cosi{
+		suite:   suite,
+		private: private,
+		publics: publics,
+	}
+	// Start with an all-disabled participation mask, then set it correctly
+	cos.mask = NewCosiMask(suite, publics)
+	return cos
+}
+
+// Announcement holds only the timestamp for that round
+type Announcement struct {
+	Timestamp int64
+}
+
+// Commitment sends it's own commit Vi and the aggregate children's
+// commit V^i
+type Commitment struct {
+	Commitment     abstract.Point
+	ChildrenCommit abstract.Point
+}
+
+// Challenge is the Hash of Commit || Publics || Msg
+// Commit is the aggregated global commit
+type Challenge struct {
+	Challenge        abstract.Secret
+	GlobalCommitment abstract.Point
+}
+
+// Response holds the actual node's response ri and the
+// aggregate response r^i
+type Response struct {
+	Response     abstract.Secret
+	ChildrenResp abstract.Secret
+}
+
+// CreateAnnouncement creates an Announcement message with the timestamp set
+// to the current time.
+func (c *Cosi) CreateAnnouncement() *Announcement {
+	now := time.Now().Unix()
+	c.timestamp = now
+	return &Announcement{now}
+}
+
+// Announce stores the timestamp and relays the message.
+func (c *Cosi) Announce(in *Announcement) *Announcement {
+	c.timestamp = in.Timestamp
+	return in
+}
+
+// CreateCommitment creates the commitment out of the random secret and returns
+// the message to pass up in the tree. This is typically called by the leaves.
+func (c *Cosi) CreateCommitment(r io.Reader) *Commitment {
```

Comment:
1) I wanted to make clear by the methods in this API what should you be doing, (should you be *starting* a new phase or should you just pass up/down). There's practically no code duplication since the real logic is the the `genCommit` function.
2) I'm always reluctant of these magic parameters where you can give *nil* or not: I always prefer explicit rather than implicit. Golang does not have optional parameters, and I don't think it's a good idea to try to emulate that.

I'm doing the change for using the new ed25519 for the moment. If you still think that makes sense to get only one function, then I'll do it right afterward :) (so please comment if you have time !)